Implement `BoneConstraint3D` with `CopyTransform`/`ConvertTransform`/`Aim` Modifiers

TokageItLab opened the pull request:
Contain https://github.com/godotengine/godot/pull/104560

## BoneConstraint3D

A base class that allows two bones to be specified.

![image](https://github.com/user-attachments/assets/0f321690-16b7-4314-a174-7ed43556ea1d)

Note that interpolation by `SkeletonModifier3D.influence` for the entire skeleton is done for each SkeletonModifier3D.

However, constraint settings are often applied to multiple bones at the same time. Therefore, it prevents unnecessary interpolation by having the settings in an array so that the number of SkeletonModifier3Ds in the scene does not grow unnecessarily. For this reason, the BoneConstraint3D settings have the apply amount of modification to the bone for each element in the settings array.

## AimModifier3D

This is a simple version of LookAtModifier3D that only allows bone to the target without advanced options such as angle limitation or time-based interpolation. The feature is simplified, but instead it is implemented with smooth tracking without euler,

![image](https://github.com/user-attachments/assets/b4463067-d71b-4b5c-a1d1-c234602961d9)

- It depends on the quaternion constructor with arcs from the two Vector3s, so it is recommended that merge  https://github.com/godotengine/godot/pull/100991 first.

## CopyTransfromModifier3D

Apply the copied transform of the target bone to the apply bone with processing it with some masks and options.

![image](https://github.com/user-attachments/assets/cf88de7d-daad-480f-8c8b-0c2b98629580)

There are 4 ways to apply the transform, depending on the combination of `Relative` and `Additive` options.

- Relative + Additive
  - Extract target pose relative to the rest and add it to the apply bone's pose.
- Relative + Not Additive
  - Extract target pose relative to the rest and add it to the apply bone's rest.
- Not Relative + Additive
  - Extract target pose absolutely and add it to the apply bone's pose.
- Not Relative + Not Additive
  - Extract target pose absolutely and the apply bone's pose is replaced with it.

## ConvertTransfromModifier3D

Apply the copied transform of the target bone to the apply bone about the specific axis with remapping it with some options.

![image](https://github.com/user-attachments/assets/62d0d831-cdf1-4e99-b7c3-af6bade062f7)

---

### Compatibility with third-parties

I assume there is some compatibility, but cannot guarantee that the behavior is exactly the same in a particular situation (since Since it has not yet been tested/verified).

#### Blender

CopyXXX (for rotation with axis mask, rotation mode is roll only) -> CopyTransfromModifier3D
Transformation (for rotation, rotation mode is roll only) -> ConvertTransfromModifier3D
Track/LockedTrack -> AimModifier3D
DampedTrack ->AimModifier3D

#### VRMConstraint

https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_node_constraint-1.0

RotationConstraint -> CopyTransfromModifier3D with only rotation with all axes without invert flags
RollConstraint -> ConvertTransfromModifier3D with rotation mode
AimConstraint -> AimModifier3D

TokageItLab committed:
Implement BoneConstraint3D with Copy/ConvertTransfrom & TrackBone mods

fire commented:
Do we have test scenes?

TokageItLab commented:
@fire I used this model in quick check to show axes where everything is a sibling of root:

![image](https://github.com/user-attachments/assets/daee8234-d26d-4513-ad62-e9cdbf308e5f)

[constraint_demo.zip](https://github.com/user-attachments/files/18308949/constraint_demo.zip)

TokageItLab commented:
@fire For now, since the feature freeze is nearing, I separated the PR only from the refactoring part, which will cause break compatibility; Since LookAtModifier is added in 4.4, if this PR is carried over to 4.5 it will occur.

https://github.com/godotengine/godot/pull/101148

TokageItLab commented:
I will write a CI test later, because this module does not depend on external factors such as time, but purely mathematical transformations, so it is possible to check for them.

aaronfranke commented:
For reference, see also this GDScript implementation of VRM constraints I made for godot-vrm: https://github.com/V-Sekai/godot-vrm/blob/master/addons/vrm/node_constraint/bone_node_constraint.gd

Also, I made an example model here: https://github.com/vrm-c/vrm-specification/pull/444 Before merging this PR, we should write a draft PR to godot-vrm that uses this PR and test that it works correctly.

This PR looks much more comprehensive than what I wrote and seems to handle everything in VRM and more, awesome! Feel free to request a review from me once the PR is ready to go.

TokageItLab commented:
I omitted the use euler option from Copy/ConvertTransformModifier.

Since the bone pose only stores the quaternion, there is no guarantee that the euler will be stored in the next frame even if the SkeletonModifier sets the euler. Then I concluded that it would be difficult to accomplish translate euler rotation without a more fundamental implementation in the Skeleton3D class.

BTW, this is not a problem in cases where processing is performed on all three axes. Where this becomes a problem is when using ConvertTransformModifier to handle Rotation, or CopyTransformModifier to mask or invert on one or two axes.

For now, to ensure mathematical correctness, they behave as if they were copying/cancelling a roll in a particular axis or inverting an element in a Quaternion. Perhaps similar to the roll+swing rotation mode in Blender.

It may seem somewhat less intuitive, but I have explained conceptually in the documentation how they work. Also, their mathematical validation is ensured by the added unit tests.

AThousandShips commented:
Please try to not push so often, this was pushed 13 times in just 3 hours, easily clogs CI

fire commented:
Should we store in swing twist decomposed qusternions/basis?

TokageItLab commented:
> Should we store in swing twist decomposed qusternions/basis?

@fire I don't think Skeleton3D needs to retain that. However, it could be argued that these functions like `get_roll_angle()` should be made public.

AThousandShips commented:
Every time you push, no matter if you have tests, it runs CI completely here, so I'd suggest creating a separate work branch if you want to run CI repeatedly just on your own fork

fire commented:
@TokageItLab feel free to use the https://github.com/V-Sekai/godot/ or you can use your own branch under https://github.com/TokageItLab/godot

TokageItLab commented:
Fixed several bugs discovered while applying Blender's constrained settings for actual use:

- Fixed a problem with inverse_lerp causing division by zero when target's range in ConvertTransformModifier is 0
- Fixed a problem with incorrect rest space in AimModifier (this PR includes https://github.com/godotengine/godot/pull/104217)
- Fixed an issue with incorrect conversion of AimModifier's without-euler mode

Twist without constraint:
![image](https://github.com/user-attachments/assets/cdb6d448-626f-49f9-8203-1a2883640fea)

Twist with constraint:
![image](https://github.com/user-attachments/assets/aadb3ff6-ff42-49ed-a1c9-e099de92ae6a)

TokageItLab commented:
Improved handling of zero division in ConvertTransformModifier. Also, "target bone" has been renamed to "reference bone".

While "target" in Aim is clear, the word "target" in Convert and Copy is a bit confusing (although Blender consistently calls it "target"). This rename would make Aim "target" to "reference", but I think it's still understandable.

If anyone has any other better names, suggestions are welcome.

fire commented:
It's april, my time to work on this is now.

lyuma reviewed (commented):
We discussed in the animation meeting and we had consensus that this feature is good.

We want to try a few extra test cases such as the VRM constraint test models, and it needs a more thorough code review

TokageItLab commented:
I have already shared this with the team, but I will put a demo project for compatibility with VRM (by manually porting settings).

![image](https://github.com/user-attachments/assets/ba435fc5-6d78-4e85-b0c6-a86f53773f25)

[constraint_demo.zip](https://github.com/user-attachments/files/20373907/constraint_demo.zip)

Model:
https://hub.vroid.com/characters/3609285562418162352/models/5465162707471684141

Animation:
https://booth.pm/ja/items/5512385

Finger settings are not ported. There are too many, but I decide that the few files animating the fingers are sufficient without them for testing. If you find anything odd, please compare Unity's UniVRM and Godot's Modifier settings. There may be a mistake in seeing the amount or something else depending on my eyes.

TokageItLab commented:
Rebased to follow https://github.com/godotengine/godot/pull/106622#issuecomment-2920399401.

SaracenOne reviewed (approved):
I think this is okay to merge now 👍. As you pointed out, it seems to address a pretty substantial amount of features available in both Blender and the VRM spec, and seems relatively easy for the end-user to implement. Since the bone constraints operate on a list of settings, hovering a property in the property inspector does not provide a description as to what the setting does. This would be a pretty nice UX onboarding improvement, but I think this PR has sat here long enough that I'm not going to advocate holding it up any longer.

Repiteo commented:
Thanks!

beicause commented on scene/3d/aim_modifier_3d.cpp:
```diff
@@ -0,0 +1,238 @@
+}
+
+void AimModifier3D::_get_property_list(List<PropertyInfo> *p_list) const {
+	BoneConstraint3D::get_property_list(p_list);
```
It seems there is no need to call this. BoneConstraint3D's properties don't need to be validated here, and its properties will be added to parent class automatically.

beicause replied:
I'd like to remove it in #106622

TokageItLab replied:
Absolutely no. It is possible to perform validation within child classes, but in order to avoid array splitting, the properties of these child classes must be added to the property list of the parent class.

scotdoria commented:
It might be because I'm misunderstanding the system but is there a way to apply the transform in the local space of the apply bone?
